### DISCUSSION TITLE
Giving user the ability to set custom transports...

### DISCUSSION THREAD

**Discussion by cryptod00c** [wevm/wagmi#3595] [Q&A]
I'm attempting to give the user the ability to provide their own custom transports, and then obviously subsequently use them throughout the wagmi config.

it is not happy when I attempt to pass in a newConfig object to the WagmiProvider, and even here if I write over the current transports property it still isn't happy - I'm sure I'm missing a step and need to call something to refresh, reinitialize the transports, but have poured over the code and docs and can't find anything.

Is this possible, or should I attempt to solve this feature another way?

```tsx
"use client";

import { createContext, useContext, useEffect, useState } from "react";
import { env } from "@/env.mjs";
import { getDefaultConfig } from "@rainbow-me/rainbowkit";
import {
  argentWallet,
  coinbaseWallet,
  injectedWallet,
  ledgerWallet,
  metaMaskWallet,
  rabbyWallet,
  rainbowWallet,
  trustWallet,
  walletConnectWallet,
} from "@rainbow-me/rainbowkit/wallets";
import type { Config } from "@wagmi/core";
import { fallback, http } from "@wagmi/core";
import { arbitrum, bsc, mainnet, optimism, polygon } from "@wagmi/core/chains";
import { WagmiProvider } from "wagmi";

import { useLocalStorage } from "@dapp/ui";
import type { ChainId } from "@dapp/utils";

const Web3Context = createContext<any>({});

type Transports = {
  [chainId in ChainId]?: string[];
};

const defaultTransports = {
  [mainnet.id]: [
    `https://eth-mainnet.g.alchemy.com/v2/${env.NEXT_PUBLIC_ALCHEMY_ID}`,
  ],
  [bsc.id]: [`https://bsc-dataseed.binance.org/`],
  [polygon.id]: [
    `https://polygon-mainnet.g.alchemy.com/v2/${env.NEXT_PUBLIC_ALCHEMY_ID}`,
  ],
  [optimism.id]: [
    `https://optimism-mainnet.g.alchemy.com/v2/${env.NEXT_PUBLIC_ALCHEMY_ID}`,
  ],
  [arbitrum.id]: [
    `https://arbitrum-mainnet.g.alchemy.com/v2/${env.NEXT_PUBLIC_ALCHEMY_ID}`,
  ],
};

export const useWeb3Config = () => {
  const { wagmiConfig, updateTransports } = useContext<any>(Web3Context);

  return {
    wagmiConfig,
    updateTransports,
  };
};

export const Web3Provider = ({ children }) => {
  const [transports, setTransports] = useLocalStorage<Transports>(
    "custom.transports",
    defaultTransports,
  );
  const [wagmiConfig, setWagmiConfig] = useState<Config>(baseConfig);

  useEffect(() => {
    setWagmiConfig(createConfigWithRpc(transports, wagmiConfig));
  }, [transports]);

  const updateTransports = (newTransports: Transports) => {
    setTransports(newTransports);
  };

  return (
    <Web3Context.Provider
      value={{
        wagmiConfig,
        updateTransports,
      }}
    >
      <WagmiProvider config={wagmiConfig}>{children}</WagmiProvider>
    </Web3Context.Provider>
  );
};

const createConfigWithRpc = (
  transports: Transports,
  currentConfig: any,
): Config => {
  const transformedTransports = Object.fromEntries(
    Object.entries(transports).map(([chainId, urls]) => [
      chainId,
      fallback((urls || []).map((url) => http(url))),
    ]),
  );

  currentConfig.transports = transformedTransports;

  return currentConfig;
};

const baseConfig = getDefaultConfig({
  appName: "dApp",
  projectId: env.NEXT_PUBLIC_WALLET_CONNECT_PROJECT_ID,
  wallets: [
    {
      groupName: "Recommended",
      wallets: [
        metaMaskWallet,
        rabbyWallet,
        walletConnectWallet,
        coinbaseWallet,
      ],
    },
    {
      groupName: "Other",
      wallets: [
        trustWallet,
        argentWallet,
        ledgerWallet,
        injectedWallet,
        rainbowWallet,
      ],
    },
  ],

  chains: [mainnet, polygon, optimism, arbitrum, bsc],
  transports: {
    [mainnet.id]: fallback([
      http(
        `https://eth-mainnet.g.alchemy.com/v2/${env.NEXT_PUBLIC_ALCHEMY_ID}`,
      ),
    ]),
    [bsc.id]: http(`https://bsc-dataseed.binance.org/`),
    [polygon.id]: http(
      `https://polygon-mainnet.g.alchemy.com/v2/${env.NEXT_PUBLIC_ALCHEMY_ID}`,
    ),
    [optimism.id]: http(
      `https://optimism-mainnet.g.alchemy.com/v2/${env.NEXT_PUBLIC_ALCHEMY_ID}`,
    ),
    [arbitrum.id]: http(
      `https://arbitrum-mainnet.g.alchemy.com/v2/${env.NEXT_PUBLIC_ALCHEMY_ID}`,
    ),
  },
  ssr: true,
});
```

**Comment by tmm** [comment]
Can you share what you are trying to do in a [TypeScript Playground](https://www.typescriptlang.org/play)? That way we can reproduce the type error and take a look.

**Comment by cryptod00c** [comment]
I got it working @tmm - thanks. when we open-source the repo I'll have you take a look at the implementation if you have a minute. y'all are awesome. thanks for all your libraries and documentation, they make building in web3 so much more enjoyable.

**Comment by cryptod00c** [comment]
@tmm put this is in a quick repo for you...

https://github.com/cryptod00c/custom-rpc

tl;dr - we want to give the user the ability to set a custom rpc for them to use within the dApp, similar to how the wallets allow you to swap out providers.

is this the best way you could see in which this allows wagmi/viem to have the proper config with the rpc/provider client.

I went through many different types of implementations for this, but this one seemed dead-simple and is working properly...

would really appreciate any recommendations or comments you might have on this implementation!
